### PR TITLE
A240 refactor mapping tool

### PR DESCRIPTION
**Pull request by emrojo**

### CODE REVIEW

File: app/javascript/react/actions.es6 [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    }
+    const cache = getState().services.cachedTaxonomies
+    if (cache) {
+      const val = cache[taxId]
+      if (val) {
+        dispatch(setManifestValue(labwareId, address, fieldName, val.scientificName, plateId))
+        return
+      }
+    }
+    return $.ajax(taxonomyServiceUrl+'/'+taxId, {
```
Comment (pjvv) [inline app/javascript/react/actions.es6]: Found where you use the service URL!

Comment (emrojo) [reply app/javascript/react/actions.es6]: Thanks, for some reason I thought this was different. Then it is still needed

File: app/javascript/react/components/labware_tabs.jsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
+import React from "react"
+import { connect } from 'react-redux'
+import StateSelectors from '../selectors'
+import { changeTab, saveTab } from '../actions'
+import classNames from 'classnames'
+
+const logName = (name) => { }
```
Comment (pjvv) [inline app/javascript/react/components/labware_tabs.jsx:7]: Is logging only enabled for debug mode?

File: app/javascript/react/components/manifest_editor.jsx [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+  )
+}
+
+const ManifestEditor = (props) => {
+  logName('ManifestEditor')
+  if (props && props.manifest) {
+    store.dispatch(loadManifest(props))
+  }
+
+  $(document.body).on('uploadManifest', (event, request) => {
```
Comment (pjvv) [inline app/javascript/react/components/manifest_editor.jsx]: Could we swap jQuery out for something else?

File: app/javascript/react/components/mapping_tool.jsx [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        <div className="form-group">
+          <label htmlFor="form-fields">Fields on Form</label>
+          <select id="form-fields" className="form-control" name="form-fields" size="8" defaultValue={props.selectedExpected||""}>
+            <ExpectedMappingOptions {...props} />
+          </select>
+        </div>
+      </div>
+      <div className="col-md-5">
+        <div className="form-group">
+          <label htmlFor="fields-from-csv">Fields from CSV</label>
```
Comment (pjvv) [inline app/javascript/react/components/mapping_tool.jsx]: Since we are allowing both CSV and Excel files, we should probably update the instructions and references in this modal dialog.

File: app/javascript/react/components/mapping_tool.jsx [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+    )
+}
+
+class MappingToolComponent extends React.Component {
+  constructor(props) {
+    super(props)
+    this.state = {
+    }
+  }
+  componentDidUpdate(){
```
Comment (pjvv) [inline app/javascript/react/components/mapping_tool.jsx]: This method looks unused.

File: app/javascript/react/constants.es6 [hunk 6]
```diff
@@ -0,0 +1,20 @@
+const C = {
```
Comment (pjvv) [inline app/javascript/react/constants.es6]: Why are some files saved as .es6 and others as .js?

File: app/javascript/react/components/mapping_tool.jsx [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+
+
+const MappingHeaderComponent = (props) => {
+  return (
+    <Modal.Header>
+      <button type="button" className="close" onClick={props.onClickClose}
+        data-dismiss="modal" aria-label="Close">
+          <span aria-hidden="true">&times;</span>
+      </button>
+      <Modal.Title>Select CSV mappings</Modal.Title>
```
Comment (pjvv) [inline app/javascript/react/components/mapping_tool.jsx:15]: Since we are allowing CSV and Excel files, we should probably update instructions and references.

File: app/javascript/react/reducers/content.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+const resetPreviousValues = (prevState, newState) => {
+  applyForAddresses(prevState, (labwareId, addressId, fieldName) => {
+    if (!found(newState, labwareId, addressId, fieldName)) {
+      setManifestValue(newState, labwareId, addressId, fieldName, "")
+    }
+  })
+  return newState
+}
+
+// const resetPreviousValues = (prevState, newState) => {
```
Comment (pjvv) [inline app/javascript/react/reducers/content.js]: This comment can be safely removed.

File: app/javascript/react/reducers/mapping.es6 [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+export default (state = {}, action) => {
+
+  switch(action.type) {
+    case C.LOAD_MANIFEST:
+      return Object.assign({}, state, action.manifest.mapping)
+
+    case C.MATCH_SELECTION:
+      let mp = Object.assign({}, state)
+      mp.matched.push({observed: action.observed, expected: action.expected})
+      mp.observed=mp.observed.filter((elem) => {return elem != action.observed})
```
Comment (pjvv) [inline app/javascript/react/reducers/mapping.es6]: Should we be using !== instead of != to prevent type coercion here?

File: app/javascript/react/reducers/mapping.es6 [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+      mp.expected=mp.expected.filter((elem) => {return elem != action.expected})
+
+      return Object.assign({rebuild: true}, state, mp, {selectedObserved: null, selectedExpected: null})
+    case C.UNMATCH:
+      let mapping = Object.assign({}, state)
+      mapping.matched = mapping.matched.filter((elem) => {
+        return !((elem.observed == action.observed) && (elem.expected == action.expected))
```
Comment (pjvv) [inline app/javascript/react/reducers/mapping.es6]: Similarly, === instead of ==?

File: app/models/manifest/provenance_state/content_accessor.rb [hunk 11]
```diff
@@ -0,0 +1,164 @@
+require 'active_support/core_ext/module/delegation'
```
Comment (pjvv) [inline app/models/manifest/provenance_state/content_accessor.rb:1]: Why is this required here and not in the other places delegate is used?

File: config/webpack/environment.js [hunk 12]
```diff
@@ -2,6 +2,8 @@ const { environment } = require('@rails/webpacker')
 const erb =  require('./loaders/erb')
 const webpack = require('webpack')
 
+//const Reception = require('./routes.js.erb')
```
Comment (pjvv) [inline config/webpack/environment.js]: Can remove this.

File: config/webpack/environment.js [hunk 13]
```diff
@@ -2,6 +2,8 @@ const { environment } = require('@rails/webpacker')
...
+
 // [PJ] add jQuery and papaparse here so that it remains available to all JS files
```
Comment (pjvv) [inline config/webpack/environment.js]: It seems we do not require papaparse anymore?

File: spec/factories/labware_types.rb [hunk 14]
```diff
@@ -1,43 +1,43 @@
 FactoryBot.define do
 
   factory :labware_type do
-    num_of_cols 12
-    num_of_rows 8
-    col_is_alpha false
-    row_is_alpha false
+    num_of_cols { 12 }
```
Comment (pjvv) [inline spec/factories/labware_types.rb:4]: Why are we defining blocks for these variables?

File: spec/features/taxon_id_finder_spec.rb [hunk 15]
```diff
@@ -80,31 +80,32 @@
...
         context 'when writing a taxonomy id in a single input' do
           it 'uses the taxonomy service and searches for the id provided', js: true do
-            fill_in('labware[1]address[A:1]fieldName[taxon_id]', with: '3')
-            expect(page).to have_selector('.has-success')
-            expect(!find_by_id('labware[1]address[A:1]fieldName[scientific_name]').value.empty?)
-              .to eq(true)
+            fill_in('labware[0]address[A:1]fieldName[taxon_id]', with: '3')
+            expect(page).to have_selector("input[value='Some specie name']")
+
+            #expect(page.find('labware[0]address[A:1]fieldName[scientific_name]')).not_to have_selector('.has-error')
```
Comment (pjvv) [inline spec/features/taxon_id_finder_spec.rb]: Can we remove this test?

File: app/javascript/react/actions.es6 (hunk 1 above)
Comment (pjvv) [inline app/javascript/react/actions.es6]: Is the taxon ID being validated on the client and in the server?

File: app/javascript/react/components/labware_tabs.jsx (hunk 2 above)
Comment (emrojo) [reply app/javascript/react/components/labware_tabs.jsx:7]: I was using it for debugging. It should not be there. I'll remove it.

File: app/javascript/react/components/manifest_editor.jsx (hunk 3 above)
Comment (emrojo) [reply app/javascript/react/components/manifest_editor.jsx]: This piece of code is not needed anymore. uploadManifest() is called from the store, not from a trigger as it was before. I'll remove it.

File: app/javascript/react/components/mapping_tool.jsx (hunk 4 above)
Comment (emrojo) [reply app/javascript/react/components/mapping_tool.jsx]: Ok

File: app/javascript/react/components/mapping_tool.jsx (hunk 5 above)
Comment (emrojo) [reply app/javascript/react/components/mapping_tool.jsx]: Yes, this component can now be just a function. I'll change it

File: app/javascript/react/constants.es6 (hunk 6 above)
Comment (emrojo) [reply app/javascript/react/constants.es6]: .es6 for javascript using ecmascript 6
.jsx for any javascript that uses React jsx in it
.js for any vanilla javascript

File: config/webpack/environment.js (hunk 12 above)
Comment (emrojo) [reply config/webpack/environment.js]: Ok

File: config/webpack/environment.js (hunk 13 above)
Comment (emrojo) [reply config/webpack/environment.js]: Ok

File: spec/factories/labware_types.rb (hunk 14 above)
Comment (emrojo) [reply spec/factories/labware_types.rb:4]: I had some problems while running the tests that for some reason were solved after changing this to blocks

File: spec/features/taxon_id_finder_spec.rb (hunk 15 above)
Comment (emrojo) [reply spec/features/taxon_id_finder_spec.rb]: Do you mean the line or the whole test?

File: app/javascript/react/components/labware_tabs.jsx (hunk 2 above)
Comment (pjvv) [reply app/javascript/react/components/labware_tabs.jsx:7]: I think it's good to keep it for debugging and developing, maybe including a flag that will only show it when not in production?

File: app/javascript/react/actions.es6 (hunk 1 above)
Comment (emrojo) [reply app/javascript/react/actions.es6]: Yes, it seems.

File: app/javascript/react/components/manifest_editor.jsx (hunk 3 above)
Comment (pjvv) [reply app/javascript/react/components/manifest_editor.jsx]: Great, less jQuery! :D

File: app/javascript/react/constants.es6 (hunk 6 above)
Comment (pjvv) [reply app/javascript/react/constants.es6]: Unless webpack or a transpiler handles .js and .es6 files differently, won't it be less confusing to simply call them both .js files?

File: spec/factories/labware_types.rb (hunk 14 above)
Comment (pjvv) [reply spec/factories/labware_types.rb:4]: Haha, weird!

File: spec/features/taxon_id_finder_spec.rb (hunk 15 above)
Comment (pjvv) [reply spec/features/taxon_id_finder_spec.rb]: Just the line..

File: app/javascript/react/constants.es6 (hunk 6 above)
Comment (emrojo) [reply app/javascript/react/constants.es6]: I thought it gives a bit more of information for knowing which type of syntax you have to follow in the file when you open it to do some changes, but I don't have any strong opinion about it. If you think it is confusing to have this type of differentiation I'll change it.

File: spec/factories/labware_types.rb (hunk 14 above)
Comment (emrojo) [reply spec/factories/labware_types.rb:4]: It may be possible that after updating some libraries it was in a version of factorybot that deprecated this (possibly as they say here  https://robots.thoughtbot.com/deprecating-static-attributes-in-factory_bot-4-11).  
It was in commit: 4f264c8bef251a5ee5930ebb8d6e645f1f58786a

Comment (pjvv) [reply spec/factories/labware_types.rb:4]: 👍

File: app/javascript/react/constants.es6 (hunk 6 above)
Comment (pjvv) [reply app/javascript/react/constants.es6]: I see. I think it's safe to assume that we should use es6 syntax throughout and if we need to add babel at a later stage it would still be worth it?

Comment (emrojo) [reply app/javascript/react/constants.es6]: Ok

**Comment by pjvv** [comment]
If we are using yarn we should and safely can remove the `package-lock.json` file. Would you agree?

**Comment by emrojo** [comment]
Ok, removing package-lock.json

File: app/javascript/react/components/labware_content_input.jsx [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+    super(props)
+    this.state = {
+      value: props.selectedValue,
+      stateValueSelected: 'redux'
+    }
+    this.onChangeProcessThrottelledCall = this.buildThrottelledCall()
+  }
+
+  setStateValueSelected(str) {
+    console.log('set value to '+str)
```
Comment (pjvv) [inline app/javascript/react/components/labware_content_input.jsx]: The logging in this file should be wrapped in a debug/devel environment checker or removed.

**Comment by pjvv** [comment]
When trying to cancel a manifest, the error `route matches [DELETE] "/manifests/1"` is shown. Should the delete URL also be sent along similarly to the save manifest?

**Comment by pjvv** [comment]
When uploading a file, the URL also needs to be corrected.
